src/vidpf.rs

```diff
@@ -41,6 +41,107 @@ pub struct EvalState {
 }
 
+pub(crate) struct VidpfEvalNode {
+    /// String representation of the path to this node.
+    bit_str: String,
+
+    /// The word for this node. Set for every node except the root.
+    word_share: Option<Vec<Field128>>,
+
+    /// The patch check component for this node. Set for every node except the root and the leaves.
+    path_check: Option<Vec<Field128>>,
+
+    /// The node state.
+    state: EvalState,
+
+    /// Left child.
+    l: Option<Box<VidpfEvalNode>>,
+
+    /// Right child.
+    r: Option<Box<VidpfEvalNode>>,
+}
+
+impl VidpfEvalNode {
+    pub(crate) fn new_root_from_key(key: &VidpfKey) -> Self {
+        Self {
+            bit_str: "".into(),
+            word_share: None,
+            path_check: None,
+            state: key.eval_init(),
+            l: None,
+            r: None,
+        }
+    }
+
+    fn next(&mut self, bit: bool, key: &VidpfKey, input_len: usize) -> Self {
+        let bit_str = self.bit_str.clone() + if bit { "1" } else { "0" };
+        let (state, word_share) = key.eval_bit(&self.state, bit, &bit_str, input_len);
+        Self {
+            bit_str,
+            word_share: Some(word_share),
+            path_check: None,
+            state,
+            l: None,
+            r: None,
+        }
+    }
+
+    pub(crate) fn traverse(
+        &mut self,
+        path: &[bool],
+        key: &VidpfKey,
+        input_len: usize,
+        eval_proof: &mut blake3::Hasher,
+    ) -> &mut VidpfEvalNode {
+        if path.len() == 0 {
+            return self;
+        }
+
+        // If this is not the root node, then ensure the node check is initialized.
+        let mut p = if self.word_share.is_some() && self.path_check.is_none() {
+            Some(self.word_share.as_ref().unwrap().clone())
+        } else {
+            None
+        };
+
+        // Compute the left child node and update the node check.
+        if self.l.is_none() {
+            let l = self.next(false, key, input_len);
+            if let (Some(ref mut path_check), Some(ref word_share_l)) = (&mut p, &l.word_share) {
+                vec_sub(path_check, &word_share_l);
+            }
+            self.l = Some(Box::new(l));
+        }
+
+        // Compute the right child node and update the node check.
+        if self.r.is_none() {
+            let r = self.next(true, key, input_len);
+            if let (Some(ref mut path_check), Some(ref word_share_r)) = (&mut p, &r.word_share) {
+                vec_sub(path_check, &word_share_r);
+            }
+            self.r = Some(Box::new(r));
+        }
+
+        // Finish path check.
+        if let Some(mut path_check) = p {
+            if key.key_idx {
+                vec_neg(&mut path_check)
+            }
+            for x in &path_check {
+                eval_proof.update(&x.get_encoded());
+            }
+            self.path_check = Some(path_check);
+        }
+
+        if !path[0] {
+            self.l.as_deref_mut().unwrap()
```

Comment:
why do we need as_deref_mut() here?

Reply:
we have a `&mut Option<T>` and we need a `Option<&mut T>`.